README.md
# Nano FrontEnd

[[版本历史/ChangeLog](CHANGELOG.md)]

[English Version](#introduce)

### 简介

FrontEnd是Nano项目自带的Web管理门户，对集群进行图形化管理。FrontEnd的服务基于[Core模块](https://github.com/project-nano/core)的API接口实现，独立进行账号管理。

由于涉及网络配置，建议使用专用Installer进行部署，项目最新版本请访问[此地址](https://github.com/project-nano/releases)

[项目官网](https://nanos.cloud/)

[项目全部源代码](https://github.com/project-nano)

[Portal/页面源文件](https://github.com/project-nano/portal)

### 编译

环境要求

- CentOS 7 x86
- Golang 1.20

```
准备依赖的framework
$git clone https://github.com/project-nano/framework.git

准备编译源代码
$git clone https://github.com/project-nano/frontend.git

编译
$cd frontend
$go build
```

编译成功在当前目录生成二进制文件frontend

### 使用

环境要求

- CentOS 7 x86



```
执行以下指令，启动FrontEnd模块
$./frontend start

也可以使用绝对地址调用或者写入开机启动脚本，比如
$/opt/nano/frontend/frontend start

```

模块启动后会输出提供形如http://192.168.5.3:5870的Web访问地址，在浏览器中打开即可。

模块运行日志输入在log/frontend.log文件中，用于查错和调试



页面初次启动，会提示创建超级用户账号，初始化之后就能开始对Nano集群进行管理了。

!!! **请注意：FrontEnd没有密码重置功能或者特殊后门，请牢记管理员密码，遗失后将无法访问系统**!!!

此外，除了模块启动功能，FrontEnd还支持以下命令参数启动

| 命令名 | 说明                               |
| ------ | ---------------------------------- |
| start  | 启动服务                           |
| stop   | 停止服务                           |
| status | 检查当前服务状态                   |
| halt   | 强行中止服务（用于服务异常时重启） |



### 配置

模块关键配置信息存储在'config/frontend.cfg'

| 参数             | 值类型 | 默认值                                      | 说明                                                         |
| ---------------- | ------ | ------------------------------------------- | ------------------------------------------------------------ |
| **address**      | 字符串 |                                             | 提供管理页面服务的主机地址，IPv4格式                         |
| **port**         | 整数   | 5870                                        | 提供管理页面服务的主机端口，默认5870                         |
| **service_host** | 字符串 |                                             | Core模块API服务的主机地址，需要与Core模块配置一致            |
| **service_port** | 整数   | 5850                                        | Core模块API服务的监听端口，需要与Core模块配置一致            |
| **api_key**      | 字符串 | ‘ThisIsAKeyPlaceHolder_ChangeToYourContent’ | 用于Core模块API服务校验的密文，需要与Core模块配置一致        |
| **api_id**       | 字符串 | ‘dummyID’                                   | 用于Core模块API服务校验的标识ID，需要与Core模块配置一致      |
| **web_root**     | 字符串 | ‘web_root’                                  | [Portal项目](https://github.com/project-nano/portal)生成的页面文件存放路径 |

假设FrontEnd模块地址为192.168.1.167，Core模块地址192.168.1.168，示例配置文件如下

```json
{
 "address": "192.168.1.167",
 "port": 5870,
 "service_host": "192.168.1.168",
 "service_port": 5850,
 "api_key": "ThisIsAKeyPlaceHolder_ChangeToYourContent",
 "api_id": "dummyID",
 "web_root": "web_root"
}
```



### 目录结构

模块主要目录和文件如下

| 目录/文件           | 说明                     |
| ------------------- | ------------------------ |
| frontend            | 模块二进制执行文件       |
| config/frontend.cfg | 模块配置文件             |
| data/log            | 管理页面操作日志存储文件 |
| log/frontend.log    | 模块运行日志             |
| web_root            | 页面文件存放目录         |
| web_root/index.html | 页面入口                 |



### Introduce

FrontEnd is the web portal that comes with the Nano project, managing clusters via GUI. The service is implemented based on the API interfaces of the [Core module](https://github.com/project-nano/core),  with its own accounts system.

It is recommended to use a dedicated Installer for deployment. For the latest project version, please visit [this address](https://github.com/project-nano/releases).

[Official Project Website](https://us.nanos.cloud/en/)

[Full Source Code of the Project](https://github.com/project-nano)

[Portal page source code](https://github.com/project-nano/portal)

### Compilation

Requirements

- CentOS 7 x86
- Golang 1.20

```
Prepare the dependent framework
$git clone https://github.com/project-nano/framework.git

Prepare the source code for compilation
$git clone https://github.com/project-nano/frontend.git

Compile
$cd frontend
$go build
```

The compiled binary file "frontend" will be generated in the current directory when success.

### Usage

Requirements

- CentOS 7 x86

```
Start module
$./frontend start

Alternatively, you can use an absolute address or write it into a startup script, such as:
$/opt/nano/frontend/frontend start
```

After the module starts, it will provide a web address that can be accessed via browser, such as http://192.168.5.3:5870.

The module logs are output to file: log/frontend.log

**!!! Please note: There is no password recover or backdoor in FrontEnd. Please remember your password carefully, or it will lost forever. !!!** 

FrontEnd also supports the following commands:

| Command | Description                                                  |
| ------- | ------------------------------------------------------------ |
| start   | Start the service                                            |
| stop    | Stop the service                                             |
| status  | Check the current service status                             |
| halt    | Forcefully terminate the service (restart when there is an exception) |

### Configuration

The main configuration is stored in file: config/frontend.cfg

| Parameter        | Value Type | Default Value                               | Description                                                  |
| ---------------- | ---------- | ------------------------------------------- | ------------------------------------------------------------ |
| **address**      | String     |                                             | The host address for the management portal, IPv4 such as '192.168.3.1' |
| **port**         | Integer    | 5870                                        | The host port for the management portal, default is 5870     |
| **service_host** | String     |                                             | The host address of API service of Core module. Must be the same as the configuration of Core module |
| **service_port** | Integer    | 5850                                        | The listening port of API service of Core module. Must be the same as the configuration of Core module. default is 5850 |
| **api_key**      | String     | 'ThisIsAKeyPlaceHolder_ChangeToYourContent' | The encryption text used for verifying API service, must be the same as the configuration of Core module. |
| **api_id**       | String     | 'dummyID'                                   | The ID used for verifying API service, Must be the same as the configuration of Core module. |
| **web_root**     | String     | 'web_root'                                  | [Portal](https://github.com/project-nano/portal) generated page files |

Assuming that the FrontEnd module address is 192.168.1.167 and the Core module address is 192.168.1.168, an example configuration file is as follows:

```json
{
 "address": "192.168.1.167",
 "port": 5870,
 "service_host": "192.168.1.168",
 "service_port": 5850,
 "api_key": "ThisIsAKeyPlaceHolder_ChangeToYourContent",
 "api_id": "dummyID",
 "web_root": "web_root"
}
```

### Directory Structure

| Directory/File      | Description                         |
| ------------------- | ----------------------------------- |
| frontend            | Binary execution file of the module |
| config/frontend.cfg | Configuration file                  |
| data/log            | Web operation logs                  |
| log/frontend.log    | Module running log                  |
| web_root            | Web page files                      |
| web_root/index.html | Web entry point                     |



CHANGELOG.md
# Change Log

## [1.4.0] - 2023-11-07

### 新增

- 配置文件"frontend.cfg"新增选项max_cores用于设置创建云主机的最大核心数，默认值24
- 配置文件"frontend.cfg"新增选项max_memory用于设置创建云主机的最大内存容量(GB)，默认值32
- 配置文件"frontend.cfg"新增选项max_disk用于设置创建云主机的最大磁盘容量(GB)，默认值64

### 变更

- 更新依赖包及go版本
- 新增front_service测试用例
- 接口'GET: /system/'返回资源配额

### Added

- Add option "max_cores" to configure file "frontend.cfg", default value is 24
- Add option "max_memory" in GB to configure file "frontend.cfg", default value is 32
- Add option "max_disk" in GB to configure file "frontend.cfg", default value is 64

### Changed

- update to go 1.20
- update dependent packages
- Add test cases to front_service
- Return max resource limit when get system status on path 'GET: /system/'

### Fixed

- Search guests not filter by owner and group

## [1.3.1] - 2021-02-18

### Added

- Search guests
- Enable/Disable auto start option

## [1.3.0] - 2020-11-08

### Added

- Forward request for sync images/security policy group

### Fixed

- Return an explicit error when login with a user does not belong to any group
- Remove from the group when deleting a user

## [1.2.0] - 2020-04-15

### Added

- Query/Change storage path
- Manage system templates
- Create guest using template
- Reset monitor secret

### Fixed

- Huge memory occupied when uploading images cause OOM kill

## [1.1.3] - 2019-12-29

### Added

- Add web_root option to "frontend.cfg" for hosting portal files
- Add cors_enable option to "frontend.cfg" for CORS control

### Changed

- A new portal completely was rewritten using React
- Web pages move from 'resource' to 'web_root'

### Fixed

- Visibility interface broken
- Query logs missing and wrong order    

## [1.1.2] - 2019-12-13

### Changed

- Enable CORS request to monitor channel

## [1.1.0] - 2019-11-05

### Added 

- Add signature when call Core API
- Add go mod

### Changed

- Call core API via prefix '/api/v1/'
- Change "/media_image_files/:id" to "/media_images/:id/file/"
- Change "/disk_image_files/:id" to "/disk_images/:id/file/"

## [1.0.0] - 2019-07-14

### Added

- Create guests with QoS options

- Modify CPU priority in guest detail

- Set threshold of Disk/Network IO in guest detail

- Batch stop guests

- Enable get/update group visibility

- Record failed login

### Changed

- URL of guest operates change from '/guest/' to '/guests/'.

- Search guest/media/disk images via session

## [0.9.1] - 2019-05-27

### Added

- Get media image

- Query/Add/Remove operate log

- Add version and online manual link to footer

- Add new API "GET /media_image_search/" for filtering media images by owner and group

- Modify disk/media images

- Add operate logs in most pages

### Changed

- Default landing page change to 'login.html'

- Add group info in session

- Bind resources to current user/group: Create instance/upload&build images

- Media images filtered using the current user and group

- Display the appropriate page according to the menu after login

- Add padding space for sidenav

- Move start/stop/restart/reset instance into nano.js

## [0.8.2] - 2019-04-4

### Fixed

- Prompt an invalid dialog when removing ranges from an address pool

- Logout when deleting address pool/computing pool/storage pool

- Prompt an invalid dialog when removing a cell from computing pool

- Logout when deleting a media/disk image

- Logout when modify user/password

- Logout when deleting user

- Logout when removing role/group/group member

## [0.8.1] - 2019-03-21

### Added

- Check resource path

- Enable batch creating and deleting/modify guest name/modify image info

- Batch creating/deleting guest

- Modify user password

- Modify guest name

### Changed

- Locate resource using ABS path

- Adapt to new runnable implement

- Verify guest name before submit creating request

- Navigation menu change to the sidebar

- Update chartjs to v2.8

## [0.7.1] - 2018-12-11

### Added

- Role/Group/User management

- Session management

- Invoke system initial page when no user configured

- Enable reset system

- Add "legacy system" option when create new instance

- Add uptime to dashboard.html

## [0.6.1] - 2018-11-30

### Added

- Redirect address pool/range API

- Address pool/range management page

### Changed

- Address pool option when creating/modifing compute pool

## [0.5.1] - 2018-11-3

### Added

- Multi-language support

- Enable/Disable cell

- Enable failover option in compute pool

- Migrate instance

### Changed

- Optimize console output when starting module

## [0.4.1] - 2018-10-2

### Added

- Support storage pool management

- Modify compute pool

- Get compute cell info

- Standalone google materialize icons

- Mark instance lost status when cell stop/disconnected

- Add image id in list

### Fixed

- No response to click start or insert media button when no media available.

- Auto fresh page become slower after a long time running

- Deleted instance count in the dashboard

- Output message without clear previous info in the list

## [0.3.1] - 2018-8-29

### Added

- Create time of Instance

- Create/modify time of Image

- Snapshot management: create/ delete / revert

- Insert / eject media image

### Changed

- Enable instance funtions in instance monitor page

- Change cpu/memory usage to bar chart in dashboard/instance monitor

- Open new tab for monitor/snapshots/details of instances list

## 2018-8-17

### Added

- Support choose installed module when cloning from an image

- Enable change admin password/ create new admin/ disk auto resize/ data disk mount when Cloud-Init module installed

- Add auto fresh in instances.html

### Modified

- Optimize instances/cells auto refresh, interval reduced to 5 seconds


daemon.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"github.com/project-nano/framework"
	"os"
	"path/filepath"
)

type FrontEndConfig struct {
	ListenAddress string `json:"address"`
	ListenPort    int    `json:"port"`
	ServiceHost   string `json:"service_host"`
	ServicePort   int    `json:"service_port"`
	APIKey        string `json:"api_key"`
	APIID         string `json:"api_id"`
	WebRoot       string `json:"web_root"`
	CORSEnable    bool   `json:"cors_enable,omitempty"`
	MaxCores      int    `json:"max_cores,omitempty"`
	MaxMemory     int    `json:"max_memory,omitempty"`
	MaxDisk       int    `json:"max_disk,omitempty"`
}

type MainService struct {
	frontend *FrontEndService
}

const (
	ExecuteName      = "frontend"
	ConfigFileName   = "frontend.cfg"
	ConfigPathName   = "config"
	WebRootName      = "web_root"
	DataPathName     = "data"
	DefaultMaxCores  = 24
	DefaultMaxMemory = 32
	DefaultMaxDisk   = 64
)

func (service *MainService) Start() (output string, err error) {
	if nil == service.frontend {
		err = errors.New("invalid service")
		return
	}
	if err = service.frontend.Start(); err != nil {
		return
	}
	output = fmt.Sprintf("Front-End Module %s\nCore API: %s\nNano Web Portal: http://%s\n",
		service.frontend.GetVersion(),
		service.frontend.GetBackendURL(),
		service.frontend.GetListenAddress())
	return
}

func (service *MainService) Stop() (output string, err error) {
	if nil == service.frontend {
		err = errors.New("invalid service")
		return
	}
	err = service.frontend.Stop()
	return
}

func (service *MainService) Snapshot() (output string, err error) {
	output = "hello, this is stub for snapshot"
	return
}

func generateConfigure(workingPath string) (err error) {
	const (
		DefaultPathPerm = 0740
	)
	var configPath = filepath.Join(workingPath, ConfigPathName)
	if _, err = os.Stat(configPath); os.IsNotExist(err) {
		//create path
		err = os.Mkdir(configPath, DefaultPathPerm)
		if err != nil {
			return
		}
		fmt.Printf("config path %s created\n", configPath)
	}

	var configFile = filepath.Join(configPath, ConfigFileName)
	if _, err = os.Stat(configFile); os.IsNotExist(err) {
		fmt.Println("No configures available, following instructions to generate a new one.")
		const (
			DefaultConfigPerm   = 0640
			DefaultBackEndPort  = 5850
			DefaultFrontEndPort = 5870
		)
		var defaultWebRoot = filepath.Join(workingPath, WebRootName)
		var config = FrontEndConfig{
			MaxCores:  DefaultMaxCores,
			MaxMemory: DefaultMaxMemory,
			MaxDisk:   DefaultMaxDisk,
		}
		if config.ListenAddress, err = framework.ChooseIPV4Address("Portal listen address"); err != nil {
			return
		}
		if config.ListenPort, err = framework.InputInteger("Portal listen port", DefaultFrontEndPort); err != nil {
			return
		}
		if config.ServiceHost, err = framework.InputString("Backend service address", config.ListenAddress); err != nil {
			return
		}
		if config.ServicePort, err = framework.InputInteger("Backend service port", DefaultBackEndPort); err != nil {
			return
		}
		if config.WebRoot, err = framework.InputString("Web Root Path", defaultWebRoot); err != nil {
			return
		}
		//write
		var data []byte
		data, err = json.MarshalIndent(config, "", " ")
		if err != nil {
			return err
		}
		if err = os.WriteFile(configFile, data, DefaultConfigPerm); err != nil {
			return err
		}
		fmt.Printf("default configure '%s' generated\n", configFile)
	}

	var dataPath = filepath.Join(workingPath, DataPathName)
	if _, err = os.Stat(dataPath); os.IsNotExist(err) {
		//create path
		err = os.Mkdir(dataPath, DefaultPathPerm)
		if err != nil {
			return
		}
		fmt.Printf("data path %s created\n", dataPath)
	}
	return
}

func createDaemon(workingPath string) (service framework.DaemonizedService, err error) {
	var configPath = filepath.Join(workingPath, ConfigPathName)
	var dataPath = filepath.Join(workingPath, DataPathName)
	if _, err = os.Stat(configPath); os.IsNotExist(err) {
		err = fmt.Errorf("config path %s not available", configPath)
		return nil, err
	}
	if _, err = os.Stat(dataPath); os.IsNotExist(err) {
		err = fmt.Errorf("data path %s not available", dataPath)
		return nil, err
	}
	var s = MainService{}
	s.frontend, err = CreateFrontEnd(configPath, dataPath)
	return &s, err
}

func main() {
	framework.ProcessDaemon(ExecuteName, generateConfigure, createDaemon)
}
